src/emscatter.jl

```diff
@@ -0,0 +1,296 @@
+L = 4.0          # Width of the area
+H = 6.0          # Height of the area
+xc = [0 -1.0]    # Center of the cylinder
+r = 1.0          # Radius of the cylinder
+d_pml = 0.8      # Thickness of the PML   
+# Physical parameters 
+k = 2*π/λ        # Wave number 
+ϵ1 = 3.0         # Relative electric permittivity for cylinder
+
+# # Discrete Model
+# 
+# We import the model from the `geometry.msh` mesh file using the `GmshDiscreteModel` function defined in `GridapGmsh`. The mesh file is created with the GMSH (see the file ../assets/emscatter/MeshGenerator.jl). Note that this mesh file already contains periodic boundary information for the left and right side, and that is enough for gridap to implement the periodic boundary condition. 
+# 
+
+model = GmshDiscreteModel("../models/geometry.msh")
+
+# # FE spaces
+# 
+# We use the first-order lagrangian as the finite element function space basis. The dirihlet edges are labeld with `DirichletEdges` in the mesh file. Since our problem involves complex numbers (because of PML), we need to assign the `vector_type` to be `Vector{ComplexF64}`.
+# 
+
+# Test and trial finite element function space
+order = 1
+reffe = ReferenceFE(lagrangian,Float64,order)
+V = TestFESpace(model,reffe,dirichlet_tags=["DirichletEdges"],vector_type=Vector{ComplexF64})
+U = TrialFESpace(V,[0])
+
+# # Numerical integration
+# 
+# We generate the triangulation and a second-order Gaussian quadrature for the numerial integration. Note that we create a boundary triangulation from a `Source` tag for the line excitation. Generally, we do not need such additional mesh tags for the source, we can use a delta function to approximate such line source excitation. However, by generating a line mesh, we can increase the accuracy of this source excitation.
+# 
+
+# Generate triangulation and quadrature from model 
+degree = 2
+Ω = Triangulation(model)
+dΩ = Measure(Ω,degree)
+# Source triangulation
+Γ = BoundaryTriangulation(model;tags="Source")
+dΓ = Measure(Γ,degree)
+
+# # PML formulation
+# 
+# More detailed demonstration of the PML formulation can be found in Ref [1]. Here we first define a `s_PML` function: $s(x)=1+\mathrm{i}\sigma(u)/\omega,$ and its derivative `ds_PML`. The parameter `LH` indicates the size of the inner boundary of the PML regions. Finally, we create a function-like object `Λ` that returns the PML factors and define its derivative in gridap. 
+# 
+
+# PML parameters
+R = 1e-4      # Tolerence for PML reflection 
+σ = -3/4*log(R)/d_pml
+LH = [L,H]
+
+# PML coordinate streching functions
+function s_PML(x,σ,k,LH,d_pml)
+    u = abs.(Tuple(x)).-LH./2
+    return @. ifelse(u > 0,  1+(1im*σ/k)*(u/d_pml)^2, $(1.0+0im))
+end
+
+function ds_PML(x,σ,k,LH,d_pml)
+    u = abs.(Tuple(x)).-LH./2
+    ds = @. ifelse(u > 0, (2im*σ/k)*(1/d_pml)^2*u, $(0.0+0im))
+    return ds.*sign.(Tuple(x))
+end
+
+struct Λ<:Function
+    σ::Float64
+    k::Float64
+    LH::Vector{Float64}
+    d_pml::Float64
+end
```

Comment:
Note that here we are defining a "callable object"  of type `Λ` that encapsulates all of the PML parameters.   This is convenient, both because we can pass lots of parameters around easily and also because we can define additional methods on `Λ`, e.g. to express the `∇(Λv)` operation